### ISSUE TITLE
Generic Connector Communication Error

### ISSUE THREAD

**Issue by dmarinello** [Bilal-S/iis2tomcat#68]
I continue to get the following error when accessing web pages on my Lucee Server:

Generic Connector Communication Error:
 Please check and adjust your setup:
 Ensure that Tomcat is running on given host and port.
 If this is a timeout error consider adjusting IIS timeout by changing executionTimeout attribute in web.config (see manual).

I’ve tried modifying server.xml settings and Boncode settings as recommended.

For awhile I've had a task that resets the lucee service every hour help the situation. Now, when the I receive the error, even resetting the services doesn’t eliminate the error. I have to restart my server.

Is there anything that I can do to correct this issue? Is there another connector that can be used?
Is there an update for boncode connector that fixes this error?

Dale

**Comment by Bilal-S** [comment]
What version of the connector are you using?
Run the connector version test on localhost:
localhost/a.cfm?BonCodeConnectorVersion=true

If you are running 1.0.37 can you send me your settings.
If not please upgrade and let me know if the problem still persists.

**Comment by dmarinello** [comment]
Bilal,

 

I’m sure it’s not the current version.

 

What’s the procedure to upgrade the BonCode and maintain the current settings and root directory for my webpages?

 

I ran the 1.0.35 update on a test server and when I was done, it changed the root directory for my webpages.

 

Thank you,

 

Dale

**Comment by Bilal-S** [comment]
You are the first to report an automatic root change. 
This only happens when you ask the installer to change it. 
It is not the default. 
Here are instructions for upgrade:
http://www.boncode.net/boncode-connector/upgrading-railo-or-lucee-connector

**Comment by dmarinello** [comment]
Bilal,

 

I ran the latest boncode update and followed your instructions taking all the defaults.

 

After the installation, the root directory changes from c:\inetpub\wwwroot to c:\lucee\tomcat\webapps]\root .

 

The websites that are off of wwwroot directory won’t come up.

 

Dale

**Comment by dmarinello** [comment]
Bilal,

 

I used the exact instructions from the website you provided.

 

Maybe you can give me the specific responses that I should use for each popup instead of taking the defaults.

 

The server is a VM, so I was able to just revert back to a snapshot.

 

Thank you,

 

Dale

**Comment by Bilal-S** [comment]
Dale,
this is becoming more and more puzzling. 
Since you can record the process can you reset and record it then send it?
The instructions do not use the word `default` at all. I am copying them in here. You followed these correct?

http://www.boncode.net/boncode-connector/upgrading-railo-or-lucee-connector

```
First make a backup of the connector settings and install files. They are located in c:\windows\BonCodeAJP13.settings and [Railo/Lucee install]\AJP13 .

The actual upgrade involves removing the existing connector and then installing the new version.

a) MAKE A BACKUP: c:\windows\BonCodeAJP13.settings and [Railo/Lucee install]\AJP13 .

b) Uninstall the existing connector: Run the Uninstallation program accessible in your control panel (Program and Features). Keep IIS features when prompted.

c) Download new version package (zip file) from distribution point (http://www.boncode.net/boncode-connector)

d) Unblock zip file: (http://boncode.net/connector/webdocs/Tomcat_Connector.htm#_Toc520189772) the downloaded .zip and, then, extract files.

e) Copy the extracted files OVER the existing files in your Railo/Lucee install (e.g. [install]\AJP13 where [install] is the directory you have installed Railo\Lucee.

f) Run the connector_setup.exe program and acknowledge the warning that “ALL YOUR INPUTS WILL BE IGNORED.” If you do not see this warning do not proceed. You have not completed steps a-e correctly.

g) from your initial backup in step a) copy your original setting file back from backup overriding the one located here c:\windows\BonCodeAJP13.settings

h) Stop/Start IIS.

That’s it.

```

**Comment by dmarinello** [comment]
Bilal,

 

I’ve done this multiple times with the same result.

 

My last email included the screenshots and the selections that I made at each option.

 

Is it possible to do a remote session to my server?

 

Dale

**Comment by Bilal-S** [comment]
Dale,
I have no record of video or image from you. I'd be glad to look at it if you could send it. My email is on the last page of the documentation pdf. To attach images to your response to GitHub you will need to use the GitHub site.

Best,
Bilal

**Comment by dmarinello** [comment]
Bilal,

Please see the attached file.

Dale
[boncode.pdf](https://github.com/Bilal-S/iis2tomcat/files/2229138/boncode.pdf)

**Comment by Bilal-S** [comment]
Thanks Dale.
A few things to note:
a)
Normally an in-place upgrade when the files are moved to the correct location and the setup is triggered should display this attached dialog. I have run this again, and doing an in-place upgrade and it did display for me.

![inputignoredmodal](https://user-images.githubusercontent.com/668901/43225941-75c70ef8-9028-11e8-8c3e-de42d7c7f8cc.PNG)

This is also mentioned in the upgrade instructions step f)

b)
Despite all, it actually could be that the connector is installed and working. 
What was the port for url you used on the last screen capture (webpage)? Was that a call to port 8088 or 80? If you did not specify a port, then it would be 80.

If so , then the connector is working. A lucee error page is a perfectly valid outcome.

If not, i.e. you are not using port 8088 can you do a screen recording what you are actually doing. 

Thanks,
Bilal

**Comment by dmarinello** [comment]
Bilal,

 

Your “All Your Inputs” screen did come up.   The first several screenshots I sent were only ones where selections were required.

 

I’m using port 80.

 

I included the Error page only to show you that the default webroot had changed after the update to c:\lucee\tomcat\webapps\root instead of remaining as c:\inetpub\wwwroot .

 

Prior to the upgrade, if I was to call up page, you can see that Lucee was looking for the file in  c:\inetpub\wwwroot.

 

cid:image001.png@01D4241E.786C76A0

 

Dale

**Comment by Bilal-S** [comment]
Ok thanks.
There is simply no code anywhere in the installer or connector to make such a change. 

What is your IIS physical path value (basic settingn for site)?
What is your tomcat server.xml webdoc root?

Simply check after install. Are they the same? Have they changed?

However, if you have a context mapping issue, i.e. Tomcat cannot identify or recognizes your site, it will fall back to the default context. Then, this is what you would see since the default context for lucee is : c:\lucee\tomcat\webapps\root
You may see errors in you tomcat logs to this effect.

**Comment by dmarinello** [comment]
Bilal,

What is your IIS physical path value (basic setting for site)?  %SystemDrive%\inetpub\wwwroot
What is your tomcat server.xml webdoc root?  I don’t see any setting for Webdoc root.  What is the specific tag?

Dale

**Comment by Bilal-S** [comment]
Dale:
the lucee setting for context (similar to an IIS site) is in server.xml file. It should **not** be commented out and is nested under the `HOST` tag. The physical documents are in the `docBase` of the `Context` tag.

```
<!-- ADD NEW HOSTS HERE -->
<Host name="[ENTER DOMAIN NAME]" appBase="webapps">
	<Context path="" docBase="[ENTER SYSTEM PATH]" />
</Host>
```

If you only see commented Hosts (this is the default in Lucee 5.x), then you are attempting to use a dynamic determined Context. In order for this to work you are relying another module/valve during the request processing called `mod_cfml`:
```
<Valve className="mod_cfml.core"
	sharedKey="[the security key]">
```

If mod_cfml cannot determine your context, you will fall back on your default context. My guess would be that this is most likely what you are experiencing. There is, of course, a completely different support group for it, but I will try to assist here.

To prove:
Run the process that you do for your upgrade. Then check your IIS settings for physical path, Has it changed? Has the server.xml changed? 
How? When (at what step)?

This should help you narrow down the problem. If you believe that this happens after the connector install, then let me know. I need to be able to replicate it and we might need a screen share.

Best,
Bilal

**Comment by dmarinello** [comment]
Bilal,

 

Everything is default in the server.xml file.

 

After I run the update, nothing is changed in the server.xml .  It has its original date creation and modified dates.

 

I copy the old .settings file over the new one in in c:\windows.

 

I restart IIS.

 

At that point, my web page will come up with the root being c:\inetpub\wwwroot.

 

When I restart the Lucee service, then the root changes to c:\lucee\tomcat\webapps\root .

 

What doesn’t make sense is if everything is default before the update and the root is c:\inetpub\wwwroot and everything is default after the update and the server.xml hasn’t changed, why isn’t the root directory the same?

 

Dale

**Comment by dmarinello** [comment]
Bilal,

 

Something else I noticed with the Lucee Server Administrator.

 

Dale

 

 

Before Boncode Update:

 



 

After Boncode Update:

**Comment by Bilal-S** [comment]
Dale:
reminder that attachments are not transferred to this ticket unless you upload in GitHub. Please use the Github website to upload images.

All data so far points consistently at Tomcat context resolution issues. You are being pushed to the default context since nothing is defined. The dynamic definition process though `mod_cfml` valve is not working. 

Few questions:
a) What are your BonCodeAJP13.settings? 
b) Do you have the following boncode settings that match your server.xml in settings after upgrade:

```
<EnableHeaderDataSupport>True</EnableHeaderDataSupport>
<ModCFMLSecret>[should match your mod_cfml Valve secret in server.xml]</ModCFMLSecret>
```

c) Are you upgrading anything else, e.g. lucee server.

d) If so, what is the issue you are trying to address with connector update?


Best,
Bilal

**Comment by dmarinello** [comment]
Since nothing is defined before the update, why is the mod_cfml working perfectly?  

Only after the update, when my Tomcat config file remains the same, does the root directory change.  How else can the context change if it’s not done in the Tomcat Config file?

![d1](https://user-images.githubusercontent.com/5098774/43279100-8b500b4c-90fc-11e8-9f15-d66503b8db57.png)

![d2](https://user-images.githubusercontent.com/5098774/43279104-918603ae-90fc-11e8-8626-bc5195672d10.png)

Dale

Few questions:
a) What are your BonCodeAJP13.settings?
<Settings>
<Server>localhost</Server>
<Port>8009</Port>
<EnableRemoteAdmin>True</EnableRemoteAdmin>
<EnableHeaderDataSupport>True</EnableHeaderDataSupport>
<ForceSecureSession>False</ForceSecureSession>
<AllowEmptyHeaders>False</AllowEmptyHeaders>
<ModCFMLSecret>89888e6ba1994d5106bec54b59997944de95162262d4c88ed86f959c9cbca206</ModCFMLSecret>
</Settings>

b) Do you have the following boncode settings that match your server.xml in settings after upgrade:

Yes,  the ModCFMLSecret matches the SharedKey in the Tomcat Config file.

<EnableHeaderDataSupport>True</EnableHeaderDataSupport>
<ModCFMLSecret>[should match your mod_cfml Valve secret in server.xml]</ModCFMLSecret>

c) Are you upgrading anything else, e.g. lucee server.
No

d) If so, what is the issue you are trying to address with connector update?
Eliminating the Generic Connector Timeouts that are constantly disabling my site.

**Comment by dmarinello** [comment]
Bilal,

The post on github eliminated the tags.

Dale


a) What are your BonCodeAJP13.settings?

<Settings>

<Server>localhost</Server>

<Port>8009</Port>

<EnableRemoteAdmin>True</EnableRemoteAdmin>

<EnableHeaderDataSupport>True</EnableHeaderDataSupport>

<ForceSecureSession>False</ForceSecureSession>

<AllowEmptyHeaders>False</AllowEmptyHeaders>

<ModCFMLSecret>89888e6ba1994d5106bec54b59997944de95162262d4c88ed86f959c9cbca206</ModCFMLSecret>

</Settings>

**Comment by Bilal-S** [comment]
OK. The plot becomes clearer.
Can you check your setting location pointer after upgrade.
E.g. when you run this url  that I shared with you last week:
http://localhost/a.cfm?BonCodeConnectorVersion=true

What message do you get?
Is this where your setting file is?

Wherever it is says the setting file should be, move a copy of your Boncode settings there and restart IIS. 
Let me know. 

Best,
Bilal

**Comment by dmarinello** [comment]
Bilal, 
  


BonCodeAJP Connector Version 1.0.37 
using generated defaults. No setting file in C:\Windows\Microsoft.Net\assembly\GAC_MSIL\BonCodeAJP13\v4.0_1.0.0.0__ad590a40d40745cf\BonCodeAJP13.settings 

  

That was it. 

  

After I copied the settings file there, everything came up fine. 

  

What caused the change from c:\windows to that directory for the settings file? 

  

Thank you so much!! 

  

I never would have figured that out. 

  

Dale 

----- Original Message -----

**Comment by Bilal-S** [comment]
No problem.
Glad it worked out.
An upgrade to .net version 4.5 made the resolving of assembly location different. We moved to .net 4.5 with 1.0.37
I have corrected for this in 1.0.38

**Comment by Bilal-S** [comment]
closing
